Firebase Push Notifications.md

```diff
@@ -0,0 +1,111 @@
+# Firebase push notifications
+
+Table of contents:
+
+1. [General](#general)
+2. [Setup](#setup)
+3. [Development approaches](#development-approaches)
+4. [Choosing right development approach](#choosing-development-approach)
+
+## General
+In this chapter we won't talk about specific implementation because everything is well described in Firebase documentation. We will explain few ideas for push notifications implementations and theirs PROS and CONS.
+
+There are many solutions which are providing us push notifications. We are using Firebase because of simplicity, reliability and it's free.
+
+Firebase Cloud Messaging handles the routing and delivery to targeted devices.
+
+Largest size of one message is 4KB.
+
+Google provides connection servers for HTTP and XMPP protocols but we are using HTTP.
+It is possible to send push notification to only one device, group of devices and topics.
+
+Before development, read chapters about [Firebase Cloud Messaging](https://firebase.google.com/docs/cloud-messaging/), and [app server development](https://firebase.google.com/docs/cloud-messaging/server).
+
+## Setup
+1. DevOps has to create new Firebase project.
+
+2. Server key from Firebase project settings should be copied to Vault. Server key authorizes your app server for access to Google services. Don't include the server key anywhere in your client code.
+
+3. For development we are using [fcm gem](https://github.com/spacialdb/fcm). Fcm gem has methods for sending requests to [Firebase](https://firebase.google.com/docs/cloud-messaging/send-message).
+
+4. Choose one development approach and you can start developing.
+
+## Development Approaches
+
+Before development starts we need to know which approaches exist.
+
+We will explain three approaches:
+  * Approach based on one device id
+  * Approach based on multiple device ids
+  * Approach based on topics
+    * With persisted topic names in our DB
+    * With non persisted topic names in our DB
+
+
+1. Approach based on one device id
+ * If user can have only one device, device_id can be stored in User model. When user installs mobile application he will get device_id from Firebase.
+
+ * Create API endpoint at which mobile team can send device_id after login. After every login they have to send you device_id because user can login in different device.
+
+ * Create API endpoint at which mobile team can delete device_id after logout. After every logout from device, mobile team will delete device_id from User through API.
```

Comment:
If the user logsout, how to you know he is still authorized to delete that device id?